core/shared/src/main/scala/zio/stm/ZSTM.scala

```diff
@@ -1040,10 +1040,20 @@ object ZSTM {
    * Applies the function `f` to each element of the `Chunk[A]` and
    * returns a transactional effect that produces a new `Chunk[B]`.
    */
-  def foreach[R, E, A, B](in: Chunk[A])(f: A => ZSTM[R, E, B]): ZSTM[R, E, Chunk[B]] =
-    in.foldLeft[ZSTM[R, E, Chunk[B]]](ZSTM.succeedNow(Chunk.empty))((acc, a) =>
-      f(a).zipWith(acc)((b, acc) => acc ++ Chunk.single(b))
-    )
+  def foreach[R, E, A, B](in: Chunk[A])(f: A => ZSTM[R, E, B]): ZSTM[R, E, Chunk[B]] = {
+    val length = in.length
+    var idx    = 0
+
+    var tx: ZSTM[R, E, ChunkBuilder[B]] = ZSTM.succeedNow(ChunkBuilder.make[B])
```

Comment:
We're allocating memory in a pure function, which is a no-no. So we have to make sure this is suspended in STM. I'd add a top-level STM wrapper (`new STM`) to make sure this function becomes referentially transparent.

Reply:
Fixed via `ZSTM.suspend`.